File: qt/gen_qrubberband.cpp
// +build ignore

#include <QEvent>
#include <QMetaObject>
#include <QMoveEvent>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QPoint>
#include <QRect>
#include <QResizeEvent>
#include <QRubberBand>
#include <QShowEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOptionRubberBand>
#include <QWidget>
#include <qrubberband.h>
#include "gen_qrubberband.h"
class MiqtVirtualQRubberBand : public virtual QRubberBand {
public:
MiqtVirtualQRubberBand(Shape param1): QRubberBand(param1) {};
MiqtVirtualQRubberBand(Shape param1, QWidget* param2): QRubberBand(param1, param2) {};

virtual ~MiqtVirtualQRubberBand() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QRubberBand::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QRubberBand_MetaObject(const_cast<MiqtVirtualQRubberBand*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QRubberBand::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QRubberBand::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QRubberBand_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QRubberBand::qt_metacast(param1);

	}
};
QRubberBand* QRubberBand_new(Shape param1) {
return new MiqtVirtualQRubberBand(param1);
}
QRubberBand* QRubberBand_new2(Shape param1, QWidget* param2) {
return new MiqtVirtualQRubberBand(param1, param2);
}
void QRubberBand_virtbase(QRubberBand* src
, QWidget** outptr_QWidget
) {
*outptr_QWidget = static_cast<QWidget*>(src);
}
QMetaObject* QRubberBand_MetaObject(const QRubberBand* self) {
	return (QMetaObject*) self->metaObject();
}
void* QRubberBand_Metacast(QRubberBand* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QRubberBand_Tr(const char* s) {
	QString _ret = QRubberBand::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
Shape QRubberBand_Shape(const QRubberBand* self) {
	return self->shape();
}
void QRubberBand_SetGeometry(QRubberBand* self, QRect* r) {
	self->setGeometry(*r);
}
void QRubberBand_SetGeometry2(QRubberBand* self, int x, int y, int w, int h) {
	self->setGeometry(static_cast<int>(x), static_cast<int>(y), static_cast<int>(w), static_cast<int>(h));
}
void QRubberBand_Move(QRubberBand* self, int x, int y) {
	self->move(static_cast<int>(x), static_cast<int>(y));
}
void QRubberBand_MoveWithQPoint(QRubberBand* self, QPoint* p) {
	self->move(*p);
}
void QRubberBand_Resize(QRubberBand* self, int w, int h) {
	self->resize(static_cast<int>(w), static_cast<int>(h));
}
void QRubberBand_ResizeWithQSize(QRubberBand* self, QSize* s) {
	self->resize(*s);
}
struct miqt_string QRubberBand_Tr2(const char* s, const char* c) {
	QString _ret = QRubberBand::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QRubberBand_Tr3(const char* s, const char* c, int n) {
	QString _ret = QRubberBand::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QRubberBand_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQRubberBand*>( (QRubberBand*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QRubberBand_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQRubberBand*)(self) )->virtualbase_MetaObject();
}
void QRubberBand_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQRubberBand*>( (QRubberBand*)(self) )->handle__Metacast = slot;
}
void* QRubberBand_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQRubberBand*)(self) )->virtualbase_Metacast(param1);
}
void QRubberBand_Delete(QRubberBand* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQRubberBand*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qrubberband.go
package qt

import (
	"unsafe"
)

type QRubberBand__Shape int

const (
	QRubberBand__Line      QRubberBand__Shape = 0
	QRubberBand__Rectangle QRubberBand__Shape = 1
)

type QRubberBand struct {
	h          uintptr
	isSubclass bool
}

// NewQRubberBand constructs a new QRubberBand object.
func NewQRubberBand(param1 Shape) *QRubberBand {
	g := newQRubberBand(QRubberBand_new(param1))
	g.isSubclass = true
	return g
}

// NewQRubberBand2 constructs a new QRubberBand object.
func NewQRubberBand2(param1 Shape, param2 *QWidget) *QRubberBand {
	g := newQRubberBand(QRubberBand_new2(param1, param2.cPointer()))
	g.isSubclass = true
	return g
}

func (this *QRubberBand) MetaObject() *QMetaObject {
	return newQMetaObject(QRubberBand_MetaObject(this.h))
}

func (this *QRubberBand) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QRubberBand_Metacast(this.h, param1_Cstring))
}

func QRubberBand_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QRubberBand_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QRubberBand) Shape() Shape {
	xxxxxxxxx
}

func (this *QRubberBand) SetGeometry(r *QRect) {
	QRubberBand_SetGeometry(this.h, r.cPointer())
}

func (this *QRubberBand) SetGeometry2(x int, y int, w int, h int) {
	QRubberBand_SetGeometry2(this.h, (int)(x), (int)(y), (int)(w), (int)(h))
}

func (this *QRubberBand) Move(x int, y int) {
	QRubberBand_Move(this.h, (int)(x), (int)(y))
}

func (this *QRubberBand) MoveWithQPoint(p *QPoint) {
	QRubberBand_MoveWithQPoint(this.h, p.cPointer())
}

func (this *QRubberBand) Resize(w int, h int) {
	QRubberBand_Resize(this.h, (int)(w), (int)(h))
}

func (this *QRubberBand) ResizeWithQSize(s *QSize) {
	QRubberBand_ResizeWithQSize(this.h, s.cPointer())
}

func QRubberBand_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QRubberBand_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QRubberBand_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QRubberBand_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QRubberBand) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QRubberBand_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QRubberBand) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QRubberBand_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRubberBand_MetaObject
func miqt_exec_callback_QRubberBand_MetaObject(self QRubberBand, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QRubberBand{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QRubberBand) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QRubberBand_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QRubberBand) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QRubberBand_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRubberBand_Metacast
func miqt_exec_callback_QRubberBand_Metacast(self QRubberBand, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QRubberBand{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
